### PR TITLE
Implement Semaphore#withPermitsManaged

### LINKED ISSUE #1289: Add a version of `Semaphore#withPermit/withPermits` that returns a Managed
This would be more composable compared to `withPermit(task)`, and safer compared to bare usage of `acquire/release`.

In the implementation, it's important to put the `prepare` step in the uncancelable reservation step, and `awaitAcquire` in the cancelable step.

### PR DESCRIPTION
**Pull request by vasilmkd**
Resolves #1289.

cc @iravid

### CODE REVIEW

File: core/shared/src/main/scala/zio/Semaphore.scala [hunk 1]
```diff
@@ -96,17 +96,31 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
    * Acquires `n` permits, executes the action and releases the permits right after.
    */
   final def withPermits[R, E, A](n: Long)(task: ZIO[R, E, A]): ZIO[R, E, A] =
     prepare(n).bracket(_.release)(_.awaitAcquire *> task)
 
+  /**
+   * Acquires `n` permits in a [[zio.ZManaged]], executes the action and releases the
+   * permits in the finalizer.
+   */
+  final def withPermitsManaged[R, E, A](n: Long)(task: ZIO[R, E, A]): ZManaged[R, E, A] =
```
Comment (iravid) [inline core/shared/src/main/scala/zio/Semaphore.scala]: `task` is actually redundant here. This method should return a `ZManaged[R, E, Unit]` and the same semantics can be achieved by flatMapping the returned managed

File: core/shared/src/main/scala/zio/Semaphore.scala [hunk 2]
```diff
@@ -96,17 +96,31 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
+    ZManaged.make(prepare(n))(_.release).flatMap(a => (a.awaitAcquire *> task).toManaged_)
```
Comment (iravid) [inline core/shared/src/main/scala/zio/Semaphore.scala]: You should use the raw `ZManaged` constructor here, as you need to place the `prepare` into the ZIO that creates the reservation, the `awaitAcquire` into the Reservation's acquire and the `release` in the Reservation's release

File: core/shared/src/main/scala/zio/Semaphore.scala (hunk 1 above)
Comment (vasilmkd) [reply core/shared/src/main/scala/zio/Semaphore.scala]: Awesome, will do.

File: core/shared/src/main/scala/zio/Semaphore.scala [hunk 3]
```diff
@@ -96,17 +96,31 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
    */
   final def withPermits[R, E, A](n: Long)(task: ZIO[R, E, A]): ZIO[R, E, A] =
     prepare(n).bracket(_.release)(_.awaitAcquire *> task)
 
+  /**
+   * Acquires `n` permits in a [[zio.ZManaged]], executes the action and releases the
+   * permits in the finalizer.
+   */
+  final def withPermitsManaged[R, E](n: Long): ZManaged[R, E, Unit] =
+    ZManaged(prepare(n).map(a => Reservation(a.awaitAcquire, a.release)))
```
Comment (iravid) [inline core/shared/src/main/scala/zio/Semaphore.scala]: This is really nice.

File: core/shared/src/main/scala/zio/Semaphore.scala (hunk 2 above)
Comment (vasilmkd) [reply core/shared/src/main/scala/zio/Semaphore.scala]: Done.

File: core/shared/src/main/scala/zio/Semaphore.scala (hunk 3 above)
Comment (vasilmkd) [reply core/shared/src/main/scala/zio/Semaphore.scala]: Thanks. And thanks for all the tips.

File: core/shared/src/main/scala/zio/Semaphore.scala [hunk 4]
```diff
@@ -96,17 +96,31 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
+    withPermitsManaged(1)
 
   /**
    * Acquires `n` permits, executes the action and releases the permits right after.
    */
   final def withPermits[R, E, A](n: Long)(task: ZIO[R, E, A]): ZIO[R, E, A] =
     prepare(n).bracket(_.release)(_.awaitAcquire *> task)
 
+  /**
+   * Acquires `n` permits in a [[zio.ZManaged]], executes the action and releases the
```
Comment (iravid) [inline core/shared/src/main/scala/zio/Semaphore.scala]: Sorry could you just update the scaladoc?

Comment (vasilmkd) [reply core/shared/src/main/scala/zio/Semaphore.scala]: You're right, sorry.

Comment (vasilmkd) [reply core/shared/src/main/scala/zio/Semaphore.scala]: Done.

**Comment by vasilmkd** [comment]
Let's now try this shiny new combinator in Streams! :smile:

**Comment by iravid** [comment]
🎉

**Pull request merged**